content/post/deeplearning_in_go_part_1.md

```diff
@@ -0,0 +1,393 @@
+
+You would be right. For simple equations, having a programmer manipulatable graph object is really overkill (unless you live in Java land, where it's classes all the way down). 
+
+I would however, posit at least three advantages of having a graph object. All of these have got to do with reducing human errors.
+
+### Numerical Stability ###
+
+Consider the equation $latex y = log(1 + x)$. This equation is not [numerically stable](https://en.wikipedia.org/wiki/Numerical_stability) - for very small values of `x`, the answer will most likely be wrong. This is because of the way `float64` is designed - a `float64` does not have enough bits to be able to tell apart `1` and `1 + 10e-16`. In fact, the correct way to do $latex y = log(1 + x)$ is to use the built in library function `math.Log1p`. It can be shown in this simple program:
+
+```go
+func main() {
+	fmt.Printf("%v\n", math.Log(1.0+10e-16))
+	fmt.Printf("%v\n", math.Log1p(10e-16))
+}
+
+1.110223024625156e-15 // wrong
+9.999999999999995e-16 // correct
+```
+
+Now, of course the programmer may be well aware of this issue and opt to use `math.Log1p` when implementing neural network algorithms, but I'm sure you'd agree that having a library that automatically converts `log(1+x)` to use `math.Log1p(x)` be rather awesome? It reduces the element of human error.
+
+### Machine-Learning Specific Optimizations ###
+
+Consider a variant of the first program in this post:
+
+```go
+func a() int {
+	return 1 + 1
+}
+```
+
+This is the same program compiled down to assembly:
+
+```
+"".a t=1 size=10 args=0x8 locals=0x0
+	(main.go:5)	TEXT	"".a(SB), $0-8
+	(main.go:5)	FUNCDATA	$0, gclocals·2a5305abe05176240e61b8620e19a815(SB)
+	(main.go:5)	FUNCDATA	$1, gclocals·33cdeccccebe80329f1fdbee7f5874cb(SB)
+	(main.go:6)	MOVQ	$2, "".~r0+8(FP)
+	(main.go:6)	RET
+```
+
+In particular, pay attention to the second last line: `MOVQ	$2, "".~r0+8(FP)`. The function has been optimized in such a way that `2` is returned. No addition operation will be performed at run time. This is because the compiler knows, *at compile time*, that 1 + 1 = 2. By replacing the expression with a constant, the compiler is saving on computation cycles at run time. If you're interested in building compilers, this is known as [constant folding](https://en.wikipedia.org/wiki/Constant_folding).
+
+So, we've established that compilers are smart enough to do optimizations. But the Go compiler (and in fact most non-machine-learning specific compilers) aren't smart enough to handle values that are used for machine learning. For machine learning, we frequently use array-based values, like a slice of `float64`s, or a matrix of `float32`s. 
```

Comment:
"But the Go compiler (and in fact most non-machine-learning specific compilers) isn't smart enough ..."